core/src/main/java/com/datastax/oss/driver/internal/core/protocol/ByteBufSegmentBuilder.java

```diff
@@ -0,0 +1,174 @@
+    for (int i = 0; i < sliceCount; i++) {
+      ChannelPromise segmentPromise = context.newPromise();
+      segmentPromise.addListener(sliceListener);
+      segmentPromises.add(segmentPromise);
+    }
+    return segmentPromises;
+  }
+
+  @Override
+  protected void processSegment(
+      @NonNull Segment<ByteBuf> segment, @NonNull ChannelPromise segmentPromise) {
+    context.write(segment, segmentPromise);
+  }
+
+  @Override
+  protected void onLargeFrameSplit(@NonNull Frame frame, int frameLength, int sliceCount) {
+    LOG.trace(
+        "[{}] Frame {} is too large ({} > {}), splitting into {} segments",
+        logPrefix,
+        frame.streamId,
+        frameLength,
+        Segment.MAX_PAYLOAD_LENGTH,
+        sliceCount);
+  }
+
+  @Override
+  protected void onSegmentFull(
+      @NonNull Frame frame, int frameLength, int currentPayloadLength, int currentFrameCount) {
+    LOG.trace(
+        "[{}] Current self-contained segment is full ({}/{} bytes, {} frames), processing now",
+        logPrefix,
+        currentPayloadLength,
+        Segment.MAX_PAYLOAD_LENGTH,
+        currentFrameCount);
+  }
+
+  @Override
+  protected void onSmallFrameAdded(
+      @NonNull Frame frame, int frameLength, int currentPayloadLength, int currentFrameCount) {
+    LOG.trace(
+        "[{}] Added frame {} to current self-contained segment "
+            + "(bringing it to {}/{} bytes, {} frames)",
+        logPrefix,
+        frame.streamId,
+        currentPayloadLength,
+        Segment.MAX_PAYLOAD_LENGTH,
+        currentFrameCount);
+  }
+
+  @Override
+  protected void onLastSegmentFlushed(int currentPayloadLength, int currentFrameCount) {
+    LOG.trace(
+        "[{}] Flushing last self-contained segment ({}/{} bytes, {} frames)",
+        logPrefix,
+        currentPayloadLength,
+        Segment.MAX_PAYLOAD_LENGTH,
+        currentFrameCount);
+  }
+
+  @NotThreadSafe
+  static class SliceWriteListener implements GenericFutureListener<Future<Void>> {
+
+    private final ChannelPromise parentPromise;
+
+    // All slices are written to the same channel, and the segment is built from the Flusher which
+    // also runs on the same event loop, so we don't need synchronization.
+    private int remainingSlices;
+    private boolean alreadyFailed;
+
+    SliceWriteListener(@NonNull ChannelPromise parentPromise, int remainingSlices) {
+      this.parentPromise = parentPromise;
+      this.remainingSlices = remainingSlices;
+    }
+
+    @Override
+    public void operationComplete(@NonNull Future<Void> future) {
+      if (!alreadyFailed) {
+        if (future.isSuccess()) {
+          remainingSlices -= 1;
+          if (remainingSlices == 0) {
+            parentPromise.setSuccess();
+          }
+        } else {
+          // No need to wait for the outcome of the other slices really. Fail immediately and we'll
+          // ignore them.
+          parentPromise.setFailure(future.cause());
+          alreadyFailed = true;
```

Comment:
why do we need this `alreadyFailed` if we are setting `setFailure` and this works on the same eventLoop? Isn't there a way to just break the flow to prevent subsequent calls to `operationComplete`?

Reply:
Good point. Maybe it's possible to cancel the other promises, I'm not sure how Netty handles that. I'll try.

Reply:
Yep, Netty detects cancelled promises and doesn't write out the bytes:
```java
EmbeddedChannel channel = new EmbeddedChannel();
ChannelPromise promise1 = channel.newPromise();
channel.write(1, promise1);
ChannelPromise promise2 = channel.newPromise();
channel.write(2, promise2);
promise2.cancel(false);
channel.flush();
System.out.println(channel.outboundMessages());
// [1, EmptyByteBufBE]
```
This happens in [ChannelOutboundBuffer](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L149-L153), so it applies to all `AbstractChannel` implementations.

Not sending the remaining slices makes sense, since the other end won't be able to reassemble the segment anyway. I'll make the change.

Reply:
Done. It's still better to handle reentry, because `operationComplete` is invoked even for a cancellation, but it occurred to me that I can check the parent promise instead of using a separate boolean.